src/platforms/common/stm32/traceswoasync.c

```diff
@@ -109,6 +130,22 @@ void traceswo_setspeed(uint32_t baudrate)
 	usart_enable_rx_dma(SWO_UART);
 }
 
+#if defined(DMA_STREAM0)
+void SWO_DMA_ISR(void)
+{
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF);
+		memcpy(&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], pingpong_buf, TRACE_ENDPOINT_SIZE);
+	}
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF);
+		memcpy(
+			&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], &pingpong_buf[TRACE_ENDPOINT_SIZE], TRACE_ENDPOINT_SIZE);
+	}
+	write_index = (write_index + 1U) % NUM_TRACE_PACKETS;
+	trace_buf_drain(usbdev, TRACE_ENDPOINT | USB_REQ_TYPE_IN);
+}
+#else
```

Comment:
This isn't a request to change any of this new code, but more that it would be really nice if these two interrupt handler implementations could be merged - raw register accesses replaced by the `dma_*` calls. It would be nice not to have to maintain two copies of the same routine after all.

Reply:
I considered unifying the implementations, but also I like to code all interrupt handlers using direct CMSIS Device register accesses for performance reasons. This DMA Stream F4 implementation is harder to inline manually because of flags being scattered across `DMA_HISR/DMA_LISR` and `DMA_HIFCR/DMA_LIFCR`, depending on which stream a given platform uses (0-3, 4-7) for USART Rx. Jumping to these functions may not be expensive thanks to ART cache, but this fires once per 64 bytes. I doubt we should fork these two functions from libopencm3 into static inline impls in headers for BMD platforms.
STM32F103, however, will suffer from similar jumps and so it benefits from coded direct accesses to `DMA_ISR` and `DMA_IFCR`. GD32F103 may not care thanks to 0WS.
The generic memcpy call TBH should be replaced with a faster memcpy4 or `__aeabi_memcpy4` intrinsic for speed, both buffers are aligned by at least 4 bytes (malloc and linkerscript) and are at least 64 bytes long (packet size). But I digress.

Reply:
> but also I like to code all interrupt handlers using direct CMSIS Device register accesses for performance reasons.

This handler will be (is) dominated by the `memcpy()` calls, the difference in maintainability here outweighs any small performance losses.

> The generic memcpy call TBH should be replaced with a faster memcpy4 or __aeabi_memcpy4 intrinsic for speed

`memcpy()` is a pseudofunction, always. It will already always be being replaced by `__builtin_memcpy()` by the compiler. This doesn't change that copying 64 bytes of memory is still the heavier operation than the register pokes, but does mean that it'll already be the most performant it can be. If the compiler can reason that it can replace the builtin with what the function does as inline asm, it'll even do that.

Reply:
Thanks for reminding me that F103 may also jump to the 28-byte memcpy function, that is a significant enough technical reason for me to not bother any further. Note however that this instance of memcpy seems to get inlined/substituted, I see no function calls in objdump/puncover other than `trace_buf_drain()` (which calls `usbd_ep_write_packet()` and so on), the `dma1_channel5_isr` handler itself is only 136 bytes.
When I apply the fix, it increases to 144 and adds two get_/clear_flag function calls which were referenced by aux_serial.c anyways. So the change is no longer status quo on flash size; for maintainability reasons, I guess.

Reply:
Aye, in this instance while smaller Flash utilisation would be nice, maintainability is a more pressing goal here in our opinion. If we turn out to be wrong? We can always re-split the IRQ handler back out and optimise, so no real loss there.